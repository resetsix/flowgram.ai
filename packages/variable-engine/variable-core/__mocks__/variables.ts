/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { ASTKind, ObjectJSON, VariableDeclarationListJSON } from '../src';

export const simpleVariableList = {
  kind: ASTKind.VariableDeclarationList,
  declarations: [
    {
      type: ASTKind.String,
      key: 'string',
    },
    {
      type: ASTKind.Boolean,
      key: 'boolean',
    },
    {
      // VariableDeclarationList 的 declarations 中可以不用声明 Kind
      // kind: ASTKind.VariableDeclaration,
      type: ASTKind.Number,
      key: 'number',
    },
    {
      kind: ASTKind.VariableDeclaration,
      type: ASTKind.Integer,
      key: 'integer',
    },
    {
      kind: ASTKind.VariableDeclaration,
      type: {
        kind: ASTKind.Object,
        properties: [
          {
            key: 'key1',
            type: ASTKind.String,
            // Object 的 properties 中可以不用声明 Kind
            kind: ASTKind.Property,
          },
          {
            key: 'key2',
            type: {
              kind: ASTKind.Object,
              properties: [
                {
                  key: 'key1',
                  type: ASTKind.Number,
                },
              ],
            } as ObjectJSON,
          },
          {
            key: 'key3',
            type: {
              kind: ASTKind.Array,
              itemType: ASTKind.String,
            },
          },
          {
            key: 'key4',
            type: {
              kind: ASTKind.Array,
              items: {
                kind: ASTKind.Object,
                properties: [
                  {
                    key: 'key1',
                    type: ASTKind.Boolean,
                  },
                ],
              } as ObjectJSON,
            },
          },
        ],
      } as ObjectJSON,
      key: 'object',
    },
    {
      kind: ASTKind.VariableDeclaration,
      type: { kind: ASTKind.Map, valueType: ASTKind.Number },
      key: 'map',
    },
  ],
} as VariableDeclarationListJSON;
