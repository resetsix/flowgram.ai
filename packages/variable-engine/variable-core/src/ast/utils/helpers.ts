/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { ASTNodeJSON, ASTNodeJSONOrKind } from '../types';
import { ASTMatch } from '../match';
import { ASTNode } from '../ast-node';

export function updateChildNodeHelper(
  this: ASTNode,
  {
    getChildNode,
    updateChildNode,
    removeChildNode,
    nextJSON,
  }: {
    getChildNode: () => ASTNode | undefined;
    updateChildNode: (nextNode: ASTNode) => void;
    removeChildNode: () => void;
    nextJSON?: ASTNodeJSON;
  }
): ASTNode | undefined {
  const currNode: ASTNode | undefined = getChildNode();

  const isNewKind = currNode?.kind !== nextJSON?.kind;
  // If `nextJSON` does not pass a key value, the key value remains unchanged by default.
  const isNewKey = nextJSON?.key && nextJSON?.key !== currNode?.key;

  if (isNewKind || isNewKey) {
    // The previous node needs to be destroyed.
    if (currNode) {
      currNode.dispose();
      removeChildNode();
    }

    if (nextJSON) {
      const newNode = this.createChildNode(nextJSON);
      updateChildNode(newNode);
      this.fireChange();
      return newNode;
    } else {
      // Also trigger an update when deleting a child node directly.
      this.fireChange();
    }
  } else if (nextJSON) {
    currNode?.fromJSON(nextJSON);
  }

  return currNode;
}

export function parseTypeJsonOrKind(typeJSONOrKind?: ASTNodeJSONOrKind): ASTNodeJSON | undefined {
  return typeof typeJSONOrKind === 'string' ? { kind: typeJSONOrKind } : typeJSONOrKind;
}

// Get all children.
export function getAllChildren(ast: ASTNode): ASTNode[] {
  return [...ast.children, ...ast.children.map((_child) => getAllChildren(_child)).flat()];
}

/**
 * isMatchAST is same as ASTMatch.is
 * @param node
 * @param targetType
 * @returns
 */
export function isMatchAST<TargetASTNode extends ASTNode>(
  node?: ASTNode,
  targetType?: { kind: string; new (...args: any[]): TargetASTNode }
): node is TargetASTNode {
  return ASTMatch.is(node, targetType);
}
