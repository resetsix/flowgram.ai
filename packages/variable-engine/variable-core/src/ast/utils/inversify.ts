/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { interfaces } from 'inversify';

import { type ASTNode } from '../ast-node';

export const injectToAST = (serviceIdentifier: interfaces.ServiceIdentifier) =>
  function (target: any, propertyKey: string) {
    if (!serviceIdentifier) {
      throw new Error(
        `ServiceIdentifier ${serviceIdentifier} in @lazyInject is Empty, it might be caused by file circular dependency, please check it.`,
      );
    }

    const descriptor = {
      get() {
        const container = (this as ASTNode).scope.variableEngine.container;
        return container.get(serviceIdentifier);
      },
      set() {},
      configurable: true,
      enumerable: true,
    } as any;

    // Object.defineProperty(target, propertyKey, descriptor);

    return descriptor;
  };

export const POST_CONSTRUCT_AST_SYMBOL = Symbol('post_construct_ast');

export const postConstructAST = () => (target: any, propertyKey: string) => {
  // 只运行一次
  if (!Reflect.hasMetadata(POST_CONSTRUCT_AST_SYMBOL, target)) {
    Reflect.defineMetadata(POST_CONSTRUCT_AST_SYMBOL, propertyKey, target);
  } else {
    throw Error('Duplication Post Construct AST');
  }
};
