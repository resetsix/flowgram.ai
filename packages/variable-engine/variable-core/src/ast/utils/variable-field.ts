/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { ASTNodeFlags } from '../flags';
import { BaseVariableField } from '../declaration';
import { ASTNode } from '../ast-node';

/**
 * 父变量字段，通过由近而远的方式进行排序
 */
export function getParentFields(ast: ASTNode): BaseVariableField[] {
  let curr = ast.parent;
  const res: BaseVariableField[] = [];

  while (curr) {
    if (curr.flags & ASTNodeFlags.VariableField) {
      res.push(curr as BaseVariableField);
    }
    curr = curr.parent;
  }

  return res;
}
