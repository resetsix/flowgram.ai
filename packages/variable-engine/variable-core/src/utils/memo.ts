/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

type KeyType = string | symbol;

/**
 * 创建缓存管理器
 * @returns
 */
export const createMemo = (): {
  <T>(key: KeyType, fn: () => T): T;
  clear: (key?: KeyType) => void;
} => {
  const _memoCache = new Map<KeyType, any>();

  const memo = <T>(key: KeyType, fn: () => T): T => {
    if (_memoCache.has(key)) {
      return _memoCache.get(key) as T;
    }
    const data = fn();
    _memoCache.set(key, data);
    return data as T;
  };

  const clear = (key?: KeyType) => {
    if (key) {
      _memoCache.delete(key);
    } else {
      _memoCache.clear();
    }
  };

  memo.clear = clear;

  return memo;
};
