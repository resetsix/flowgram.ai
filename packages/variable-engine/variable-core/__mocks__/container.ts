/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { Container, ContainerModule, interfaces } from 'inversify';
import { ScopeChain, VariableContainerModule } from '../src';
import { MockScopeChain } from './mock-chain';
import { createPlaygroundContainer } from '@flowgram.ai/core';

export function getContainer(customModule?: interfaces.ContainerModuleCallBack): Container {
  const container = createPlaygroundContainer() as Container;
  container.load(VariableContainerModule);
  container.bind(ScopeChain).to(MockScopeChain).inSingletonScope();

  if(customModule) {
    container.load(new ContainerModule(customModule))
  }

  return container;
}
