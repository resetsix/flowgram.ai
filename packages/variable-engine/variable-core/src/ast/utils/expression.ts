/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { intersection } from 'lodash-es';

import { ASTNodeFlags } from '../flags';
import { type BaseExpression } from '../expression';
import { type BaseVariableField } from '../declaration';
import { type ASTNode } from '../ast-node';
import { getParentFields } from './variable-field';
import { getAllChildren } from './helpers';

// 获取所有子 AST 引用的变量
export function getAllRefs(ast: ASTNode): BaseVariableField[] {
  return getAllChildren(ast)
    .filter((_child) => _child.flags & ASTNodeFlags.Expression)
    .map((_child) => (_child as BaseExpression).refs)
    .flat()
    .filter(Boolean) as BaseVariableField[];
}

/**
 * 检测是否成环
 * @param curr 当前表达式
 * @param refNode 引用的变量节点
 * @returns 是否成环
 */
export function checkRefCycle(
  curr: BaseExpression,
  refNodes: (BaseVariableField | undefined)[]
): boolean {
  // 作用域没有成环，则不可能成环
  if (
    intersection(curr.scope.coverScopes, refNodes.map((_ref) => _ref?.scope).filter(Boolean))
      .length === 0
  ) {
    return false;
  }

  // BFS 遍历
  const visited = new Set<BaseVariableField>();
  const queue = [...refNodes];

  while (queue.length) {
    const currNode = queue.shift()!;
    visited.add(currNode);

    for (const ref of getAllRefs(currNode).filter((_ref) => !visited.has(_ref))) {
      queue.push(ref);
    }
  }

  // 引用的变量中，包含表达式的父变量，则成环
  return intersection(Array.from(visited), getParentFields(curr)).length > 0;
}
