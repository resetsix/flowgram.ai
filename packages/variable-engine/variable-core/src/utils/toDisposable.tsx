/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { Subscription } from 'rxjs';
import { Disposable } from '@flowgram.ai/utils';

export function subsToDisposable(subscription: Subscription): Disposable {
  return Disposable.create(() => subscription.unsubscribe());
}
