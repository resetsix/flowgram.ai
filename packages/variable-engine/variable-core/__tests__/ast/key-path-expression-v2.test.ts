/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { vi, describe, test, expect } from 'vitest';

import { getParentFields } from '../../src/ast/utils/variable-field';
import {
  ASTKind,
  VariableEngine,
  VariableDeclaration,
  ASTFactory,
  KeyPathExpressionV2,
} from '../../src';
import { getContainer } from '../../__mocks__/container';

const {
  createVariableDeclaration,
  createObject,
  createProperty,
  createString,
  createNumber,
  createKeyPathExpression,
  createArray,
  createEnumerateExpression,
} = ASTFactory;

vi.mock('nanoid', () => {
  let mockId = 0;
  return {
    nanoid: () => 'mocked-id-' + mockId++,
  };
});

/**
 * 测试通过表达式生成的变量
 */
describe('test Key Path Expression V2', () => {
  const container = getContainer();
  const variableEngine = container.get(VariableEngine);
  variableEngine.astRegisters.registerAST(KeyPathExpressionV2);
  const globalScope = variableEngine.createScope('global');
  const testScope = variableEngine.createScope('test');

  const globalTestVariable = globalScope.ast.set<VariableDeclaration>(
    'test',
    createVariableDeclaration({
      key: 'source',
      type: createObject({
        properties: [
          createProperty({
            key: 'a',
            type: createString(),
          }),
          createProperty({
            key: 'b',
            type: createNumber(),
          }),
        ],
      }),
    }),
  );

  test('init const test_key_path = source.a', () => {
    const variableByKeyPath = testScope.ast.set<VariableDeclaration>(
      'variableByKeyPath',
      createVariableDeclaration({
        key: 'test_key_path',
        initializer: createKeyPathExpression({
          keyPath: ['source', 'a'],
        }),
      }),
    )!;

    expect(variableByKeyPath.initializer?.kind).toEqual(ASTKind.KeyPathExpression);
    expect(variableByKeyPath.initializer?.refs.map(_ref => _ref?.key)).toEqual(['a']);
    // variableByKeyPath 的类型重新建立了实例，其父节点为当前节点
    expect(getParentFields(variableByKeyPath.type).map(_field => _field?.key)).toEqual([
      'test_key_path',
    ]);
    expect(variableByKeyPath.type.toJSON()).toEqual({ kind: ASTKind.String });
    expect(variableByKeyPath.toJSON()).toMatchSnapshot();
  });

  test('subscribe variable changes by expression', () => {
    // const variableByKeyPath = source.a;
    const variableByKeyPath = testScope.output.getVariableByKey('test_key_path')!;

    let typeChangeTimes = 0;
    variableByKeyPath.onTypeChange(type => {
      typeChangeTimes++;
      expect([typeChangeTimes, type?.toJSON()]).toMatchSnapshot();
    });

    // source.a 发生变化时，则响应更新变量
    // source.a -> boolean
    globalTestVariable.getByKeyPath(['a'])?.updateType(ASTKind.Boolean);
    expect(variableByKeyPath.type.toJSON()).toEqual({ kind: ASTKind.Boolean });
    expect(typeChangeTimes).toBe(1);

    // 改成引用 source.b，响应更新变量
    // const variableByKeyPath = source.b;
    variableByKeyPath.updateInitializer({
      kind: ASTKind.KeyPathExpression,
      keyPath: ['source', 'b'],
    });
    expect(variableByKeyPath.type.toJSON()).toEqual({ kind: ASTKind.Number });
    expect(typeChangeTimes).toBe(2);

    // source.b 被删除，变量类型变为空
    globalTestVariable.type.fromJSON({
      properties: [
        {
          key: 'a',
          type: ASTKind.String,
        },
      ],
    });

    expect(variableByKeyPath.type).toBeUndefined();
    expect(typeChangeTimes).toBe(3);

    // source.b 加回来，变量类型变回来且触发表达式更新
    globalTestVariable.type.fromJSON(
      createObject({
        properties: [
          createProperty({ key: 'a', type: createString() }),
          createProperty({ key: 'b', type: createNumber() }),
        ],
      }),
    );
    expect(variableByKeyPath.type.toJSON()).toEqual({ kind: ASTKind.Number });
    expect(typeChangeTimes).toBe(4);

    // 改成 EnumerateExpression
    variableByKeyPath.updateInitializer(
      createEnumerateExpression({
        enumerateFor: createKeyPathExpression({
          keyPath: ['source', 'b'],
        }),
      }),
    );
    expect(variableByKeyPath.type).toBeUndefined();
    expect(variableByKeyPath.initializer?.refs).toEqual([]);
    variableByKeyPath.initializer?.refreshRefs();
    expect(variableByKeyPath.initializer?.refs).toEqual([]);
    expect(typeChangeTimes).toBe(5);

    // 数组下钻，类型也会更新
    globalTestVariable.type.fromJSON(
      createObject({
        properties: [
          createProperty({
            key: 'b',
            type: createArray({
              items: createNumber(),
            }),
          }),
        ],
      }),
    );
    expect(variableByKeyPath.type.toJSON()).toEqual({ kind: ASTKind.Number });
    expect(typeChangeTimes).toBe(6);
    expect(variableByKeyPath.toJSON()).toMatchSnapshot();

    // 原作用域删除，显示为空类型
    globalScope.dispose();
    expect(variableByKeyPath.scope.depScopes.map(_scope => _scope.id)).toEqual([]);
    expect(variableByKeyPath.type).toBeUndefined();
    expect(typeChangeTimes).toBe(7);
  });

  const cycle1 = variableEngine.createScope('cycle1');
  const cycle2 = variableEngine.createScope('cycle2');
  const cycle3 = variableEngine.createScope('cycle3');

  test('cycle scope with normal refs', () => {
    const cycle1Var = cycle1.ast.set<VariableDeclaration>(
      'var',
      createVariableDeclaration({
        key: 'cycle1_var',
        type: createObject({
          properties: [
            createProperty({
              key: 'a',
              type: createArray({
                items: createString(),
              }),
            }),
          ],
        }),
      }),
    );

    const cycle2Var = cycle2.ast.set<VariableDeclaration>(
      'var',
      createVariableDeclaration({
        key: 'cycle2_var',
        initializer: createEnumerateExpression({
          enumerateFor: createKeyPathExpression({
            keyPath: ['cycle1_var', 'a'],
          }),
        }),
      }),
    );

    const cycle3Var = cycle3.ast.set<VariableDeclaration>(
      'var',
      createVariableDeclaration({
        key: 'cycle3_var',
        initializer: createKeyPathExpression({
          keyPath: ['cycle2_var'],
        }),
      }),
    );

    expect(cycle1Var.type.toJSON()).toEqual({
      kind: ASTKind.Object,
      properties: [
        {
          kind: ASTKind.Property,
          key: 'a',
          type: {
            kind: ASTKind.Array,
            items: { kind: ASTKind.String },
          },
        },
      ],
    });
    expect(cycle2Var.type.toJSON()).toEqual({
      kind: ASTKind.String,
    });
    expect(cycle3Var.type.toJSON()).toEqual({
      kind: ASTKind.String,
    });
  });

  test('cycle scope with child ref', () => {
    const cycle1Var = cycle1.ast.get('var')! as VariableDeclaration;
    const cycle2Var = cycle2.ast.get('var')! as VariableDeclaration;
    const cycle3Var = cycle3.ast.get('var')! as VariableDeclaration;

    cycle1Var.fromJSON({
      type: createObject({
        properties: [
          createProperty({
            key: 'a',
            type: createArray({
              items: createString(),
            }),
          }),
          createProperty({
            key: 'b',
            initializer: createKeyPathExpression({
              keyPath: ['cycle3_var'],
            }),
          }),
        ],
      }),
    });

    expect(cycle1Var.type.toJSON()).toEqual({
      kind: ASTKind.Object,
      properties: [
        {
          kind: ASTKind.Property,
          key: 'a',
          type: {
            kind: ASTKind.Array,
            items: { kind: ASTKind.String },
          },
        },
        {
          kind: ASTKind.Property,
          key: 'b',
          initializer: {
            kind: ASTKind.KeyPathExpression,
            keyPath: ['cycle3_var'],
          },
          type: {
            kind: ASTKind.String,
          },
        },
      ],
    });
    expect(cycle2Var.type.toJSON()).toEqual({
      kind: ASTKind.String,
    });
    expect(cycle3Var.type.toJSON()).toEqual({
      kind: ASTKind.String,
    });
  });

  test('cycle scope with cycle ref', () => {
    const cycle1Var = cycle1.ast.get('var')! as VariableDeclaration;
    const cycle2Var = cycle2.ast.get('var')! as VariableDeclaration;
    const cycle3Var = cycle3.ast.get('var')! as VariableDeclaration;

    cycle1Var.fromJSON({
      type: createObject({
        properties: [
          createProperty({
            key: 'a',
            initializer: createKeyPathExpression({
              keyPath: ['cycle3_var'],
            }),
          }),
          createProperty({
            key: 'b',
            initializer: createKeyPathExpression({
              keyPath: ['cycle3_var'],
            }),
          }),
        ],
      }),
    });

    expect(cycle1Var.type.toJSON()).toEqual({
      kind: ASTKind.Object,
      properties: [
        {
          kind: ASTKind.Property,
          key: 'a',
          initializer: {
            kind: ASTKind.KeyPathExpression,
            keyPath: ['cycle3_var'],
          },
          // 发生循环引用时，type 清空
        },
        {
          kind: ASTKind.Property,
          key: 'b',
          initializer: {
            kind: ASTKind.KeyPathExpression,
            keyPath: ['cycle3_var'],
          },
          // 发生循环引用时，type 清空
        },
      ],
    });
    expect(cycle2Var.type?.toJSON()).toBeUndefined();
    expect(cycle3Var.type?.toJSON()).toBeUndefined();
  });
});
