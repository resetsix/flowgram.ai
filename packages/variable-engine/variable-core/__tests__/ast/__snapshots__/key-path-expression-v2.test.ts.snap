// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test Key Path Expression V2 > init const test_key_path = source.a 1`] = `
{
  "initializer": {
    "keyPath": [
      "source",
      "a",
    ],
    "kind": "KeyPathExpression",
  },
  "key": "test_key_path",
  "kind": "VariableDeclaration",
  "order": 0,
  "type": {
    "kind": "String",
  },
}
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 1`] = `
[
  1,
  {
    "kind": "Boolean",
  },
]
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 2`] = `
[
  2,
  {
    "kind": "Number",
  },
]
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 3`] = `
[
  3,
  undefined,
]
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 4`] = `
[
  4,
  {
    "kind": "Number",
  },
]
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 5`] = `
[
  5,
  undefined,
]
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 6`] = `
[
  6,
  {
    "kind": "Number",
  },
]
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 7`] = `
{
  "initializer": {
    "enumerateFor": {
      "keyPath": [
        "source",
        "b",
      ],
      "kind": "KeyPathExpression",
    },
    "kind": "EnumerateExpression",
  },
  "key": "test_key_path",
  "kind": "VariableDeclaration",
  "order": 0,
  "type": {
    "kind": "Number",
  },
}
`;

exports[`test Key Path Expression V2 > subscribe variable changes by expression 8`] = `
[
  7,
  undefined,
]
`;
